File: app/src/main/java/org/xhanka/biblesiswati/MainActivity.java
package org.xhanka.biblesiswati;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.splashscreen.SplashScreen;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.navigation.NavController;
import androidx.navigation.NavOptions;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.google.android.material.navigation.NavigationView;

import org.jetbrains.annotations.NotNull;
import org.xhanka.biblesiswati.common.Utils;
import org.xhanka.biblesiswati.databinding.ActivityMainBinding;

import java.util.Objects;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class MainActivity extends AppCompatActivity {
    int idToLoad;

    private AppBarConfiguration mAppBarConfiguration;
    private NavController navController;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Utils.Companion.toggleDarkMode(this);

        super.onCreate(savedInstanceState);
        SplashScreen.installSplashScreen(this);

        ActivityMainBinding binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setSupportActionBar(binding.appBarMain.toolbar);
        DrawerLayout drawer = binding.drawerLayout;
        NavigationView navigationView = binding.navView;
        // Passing each menu ID as a set of Ids because each
        // menu should be considered as top level destinations.
        mAppBarConfiguration = new AppBarConfiguration.Builder(
                R.id.nav_home, R.id.nav_list_songs, R.id.nav_siswati_ref,
                R.id.nav_notes, R.id.nav_fav_verses, R.id.nav_about, R.id.nav_settings)
                .setOpenableLayout(drawer)
                .build();
        navController = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        NavigationUI.setupActionBarWithNavController(this, navController, mAppBarConfiguration);
        NavigationUI.setupWithNavController(navigationView, navController);

        idToLoad = 0;
        binding.navView.setNavigationItemSelectedListener(item -> {
            idToLoad = item.getItemId();
            binding.drawerLayout.closeDrawer(GravityCompat.START, true);
            return true;
        });

        binding.drawerLayout.addDrawerListener(new DrawerLayout.DrawerListener() {
            @Override
            public void onDrawerSlide(@NonNull @NotNull View drawerView, float slideOffset) {
            }

            @Override
            public void onDrawerOpened(@NonNull @NotNull View drawerView) {
            }

            @Override
            public void onDrawerClosed(@NonNull @NotNull View drawerView) {
                if (idToLoad != 0 && Objects.requireNonNull(navController.getCurrentDestination()).getId() != idToLoad) {
                    NavOptions options = new NavOptions.Builder()
                            .setLaunchSingleTop(true)
                            .setPopUpTo(navController.getGraph().getStartDestinationId(), false)
                            .setExitAnim(R.anim.nav_default_exit_anim)
                            .setEnterAnim(R.anim.nav_default_enter_anim)
                            .setPopExitAnim(R.anim.nav_default_pop_exit_anim)
                            .setPopEnterAnim(R.anim.nav_default_pop_enter_anim)
                            .build();

                    navController.navigate(idToLoad, null, options);
                }
            }

            @Override
            public void onDrawerStateChanged(int newState) {
            }
        });

        navController.addOnDestinationChangedListener((controller, navDestination, bundle) -> {
            idToLoad = navDestination.getId();
            binding.navView.setCheckedItem(navDestination.getId());
        });

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.action_settings) {
            navController.navigate(R.id.nav_settings);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onSupportNavigateUp() {
        ///NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        return NavigationUI.navigateUp(navController, mAppBarConfiguration)
                || super.onSupportNavigateUp();
    }
}

File: app/src/main/java/org/xhanka/biblesiswati/MainActivity.kt
package org.xhanka.biblesiswati

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout.DrawerListener
import androidx.navigation.NavController
import androidx.navigation.NavDestination
import androidx.navigation.NavOptions
import androidx.navigation.Navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI.setupActionBarWithNavController
import androidx.navigation.ui.NavigationUI.setupWithNavController
import androidx.navigation.ui.navigateUp
import dagger.hilt.android.AndroidEntryPoint
import org.xhanka.biblesiswati.common.Utils.Companion.toggleDarkMode
import org.xhanka.biblesiswati.databinding.ActivityMainBinding

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {

    private var _binding: ActivityMainBinding? = null
    private val binding get() = _binding!!

    var idToLoad = 0
    private lateinit var navController: NavController
    private lateinit var appBarConfiguration: AppBarConfiguration

    override fun onCreate(savedInstanceState: Bundle?) {
        installSplashScreen()
        super.onCreate(savedInstanceState)

        // always toggle dark mode after activity has initialized : called super.onCreate()
        toggleDarkMode(this)

        _binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.appBarMain.toolbar)

        val drawer = binding.drawerLayout
        val navigationView = binding.navView

        // Passing each menu ID as a set of Ids because each
        // menu should be considered as top level destinations.
        appBarConfiguration = AppBarConfiguration(
            setOf(
                R.id.nav_home, R.id.nav_list_songs, R.id.nav_siswati_ref,
                R.id.nav_notes, R.id.nav_fav_verses, R.id.nav_install_extra_versions,
                R.id.nav_about, R.id.nav_settings
            ), drawer
        )

        navController = findNavController(this, R.id.nav_host_fragment_content_main)
        setupActionBarWithNavController(this, navController, appBarConfiguration)
        setupWithNavController(navigationView, navController)

        binding.navView.setNavigationItemSelectedListener { item: MenuItem ->
            idToLoad = item.itemId
            binding.drawerLayout.closeDrawer(GravityCompat.START, true)
            true
        }

        binding.drawerLayout.addDrawerListener(object : DrawerListener {
            override fun onDrawerSlide(drawerView: View, slideOffset: Float) {}
            override fun onDrawerOpened(drawerView: View) {}
            override fun onDrawerClosed(drawerView: View) {
                if (idToLoad != 0 && navController.currentDestination?.id != idToLoad) {
                    val options: NavOptions = NavOptions.Builder()
                        .setLaunchSingleTop(true)
                        .setPopUpTo(navController.graph.startDestinationId, false)
                        .setExitAnim(R.anim.nav_default_exit_anim)
                        .setEnterAnim(R.anim.nav_default_enter_anim)
                        .setPopExitAnim(R.anim.nav_default_pop_exit_anim)
                        .setPopEnterAnim(R.anim.nav_default_pop_enter_anim)
                        .build()
                    navController.navigate(idToLoad, null, options)
                }
            }

            override fun onDrawerStateChanged(newState: Int) {}
        })
        navController.addOnDestinationChangedListener { _: NavController?, navDestination: NavDestination, _: Bundle? ->
            idToLoad = navDestination.id
            binding.navView.setCheckedItem(navDestination.id)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        ///NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        return navController.navigateUp(appBarConfiguration) || super.onSupportNavigateUp()
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }
}